Source/WebCore/page/LocalDOMWindow.cpp

```diff
@@ -2458,39 +2460,175 @@ void LocalDOMWindow::finishedLoading()
     }
 }
 
+PerformanceEventTiming::InteractionID& LocalDOMWindow::ensureUserInteractionValue()
+{
+    // User interaction value should be initialized with a random number from 100 to 10000:
+    if (m_userInteractionValue < 100) [[unlikely]]
+        m_userInteractionValue = 100 + WTF::cryptographicallyRandomNumber<uint64_t>() % 9901;
+    return m_userInteractionValue;
+}
+
+void LocalDOMWindow::increaseInteractionCount()
+{
+    ++m_interactionCount;
+    // m_userInteractionValue should be increased by a small integer greater
+    // than 1 to "discourage developers from considering it as a counter":
+    ensureUserInteractionValue() += 7;
+};
+
 PerformanceEventTiming::Candidate LocalDOMWindow::initializeEventTimingEntry(const Event& event, EventTypeInfo typeInfo)
 {
-    // FIXME: implement InteractionId logic
     auto processingStart = performance().nowInReducedResolutionSeconds();
+    PerformanceEventTiming::InteractionID interactionId = 0;
     LOG_WITH_STREAM(PerformanceTimeline, stream << "Initializing event timing entry (type=" << event.type() << ") at t=" << processingStart);
+    // Stateful interactionId logic as described in https://www.w3.org/TR/event-timing/#sec-computing-interactionid :
+    switch (typeInfo.type()) {
+    case EventType::keyup:
+        ASSERT(event.isKeyboardEvent());
+        {
+            auto keyboardEvent = downcast<KeyboardEvent>(&event);
+            ASSERT(keyboardEvent);
+            if (keyboardEvent->isComposing())
+                break;
+
+            auto code = keyboardEvent->keyCode();
+            if (!m_pendingKeyDowns.contains(code))
+                break;
+
+            increaseInteractionCount();
+            auto it = m_pendingKeyDowns.find(code);
+            it->value.interactionID = ensureUserInteractionValue();
+            m_performanceEventTimingCandidates.append(it->value);
+            m_pendingKeyDowns.remove(code);
```

Comment:
This will do 3 hash map lookups. That's awfully inefficient.
We should be doing exactly one hash map lookup using `find` and using the resultant iterator to the rest of operations.